### DISCUSSION TITLE
LinuxCNC compatabillity

### DISCUSSION THREAD

**Discussion by ShadeTechnik** [grblHAL/core#98] [General]
Not sure if this is even a reasonable request since I don't use grbl or grblHAL in anything and I don't know much about programming either. I do use LinuxCNC alot, and I use Teensy 4.x's alot with custom PCB designs. I know there's sort of an overlap in what LinuxCNC and grbl do, but this seems to have most of the blocks in place to  slip some kind of LinuxCNC HAL interface into, I'm curious what it would take to get something like that going. People have worked on Teensy's with LinuxCNC but it always seems to die since it's usually just 1 guy doing a random project.

I'm talking about something similar to what mesa does with their HM2 FPGA software. There are cards with ethernet and SPI interfaces with the stepgens and encoder modules, it'd be great to see something like that on a Teensy since it's pretty accessable, though I'm not sure if grblHAL is a means to get there. It's been discussed here:
[https://www.forum.linuxcnc.org/18-computer/43212-teensy-based-cnc-board](url)
[https://forum.linuxcnc.org/24-hal-components/39813-teensy-4-1-linuxcnc?start=20](url)

**Comment by terjeio** [comment]
You are suggesting using a Teensy instead of a Mesa card? If so then [the driver](https://github.com/grblHAL/iMXRT1062) may contain some usable code.

**Comment by ShadeTechnik** [comment]
Yeah more or less, it's been done before. Like I said those projects just didn't seem to hold much steam. I'm just curious if there's any interest from the developers of grblHAL here to support such a thing.

**Reply by terjeio** [reply]
I am sorrry to say that I am not going to do it, I have far too much on my hands already. Maybe someone else will?

**Reply by kfmut** [reply]
> projects just didn't seem to hold much steam

Because there are lacking any usable hardware implementation suitable for usage by general public so no support and no appreciation.

**Reply by ShadeTechnik** [reply]
I don't agree with that. If you look at the link I posted the guy never released the work he did to the public and he never finished it so of course that was gonna die. As for "hardware implementation" there's no reason this won't work https://www.tindie.com/products/philba/grblhal-bob-unkit-for-teensy-41-t41u5xbb/ it's even mentioned in the thread above. I don't know how many developers are on this, and alot of Mesa's open source code could be ported to this I'm sure, espescially the ethernet or SPI interface modules. Once it's done and pins are configurable in a simple manner, anyone can create a break out board.  

If it were to get done, I'll open source a PCB design. I've done several Teensy hardware boards. Everything I've made has a pretty specific purpose so I don't have anything at the moment. Software is always my biggest issue so unless someone wants to handle that part, there's no purpose in me doing anything like that. 

![20211013_124600](https://user-images.githubusercontent.com/55633786/145686661-7519829a-83c0-45f6-a774-3249d2991207.jpg)
![20200924_085845 sync-conflict-20210805-080115-BABQD45](https://user-images.githubusercontent.com/55633786/145686669-f92f1100-828f-43f3-9f2b-f46736e07177.jpg)
![20210214_161425](https://user-images.githubusercontent.com/55633786/145686699-c0e8a2ec-aed5-46e4-88ce-deea954b51c0.jpg)
![20211210_192317](https://user-images.githubusercontent.com/55633786/145686730-5e23cf97-ae1e-4141-ac65-5eb03a5431b1.jpg)
![20211210_193748](https://user-images.githubusercontent.com/55633786/145686735-7fa329b5-a072-42f5-80d7-c1ccf0bf3b2b.jpg)

**Reply by kfmut** [reply]
> anyone can create a break out board

Why they would do that if cheap mesa clones are readily available?

Correct me if I wrong, but previous "LinuxCNC for hobbyist" idea was related to SBC(rpi, opi etc)? And now it's in shambles without any usable results? Why it's so if SBCs are also "pretty accessable" and "anyone can create a break out board" for them?

**Reply by ShadeTechnik** [reply]
> > anyone can create a break out board
> 
> Why they would do that if cheap mesa clones are readily available?
I don't know about "cheap Mesa clones" I've never heard of a clone Mesa card. Your question is no different than asking "why would someone use grblHAL when they can use LinuxCNC and a Mesa card? Or why is anyone here at grblHAL messing with Teensy's when they could be developing this for FPGAs? The answer to those questions is the same as the answer to yours.

> Correct me if I wrong, but previous "LinuxCNC for hobbyist" idea was related to SBC(rpi, opi etc)? And now it's in shambles without any usable results? Why it's so if SBCs are also "pretty accessable" and "anyone can create a break out board" for them?

I never said anything about making a breakout board for an SBC, where do you see that? I said a BOB for a Teensy. The machine running LinuxCNC itself is of no concern to what I'm saying here. I'm talking about using a Teensy as the motion controller for LinuxCNC, the architecture for that is similar to grblHAL, and is what a Mesa card does, which is why I'm here. 

The project you're referring to as "LinuxCNC for Hobbyist" that is in "shambles" for "SBCs" is Machinekit  which is a fork of LinuxCNC and yes, that is damn near dead....but I never said anything about machinekit. Machinekit was great for what it did but it focused too hard on specific SBCs and fell behind LinuxCNC so far to the point that by the time LinuxCNC is working well for RPI4s, machinekit didn't even get there yet.  None of that's what I'm talking about, point is LinuxCNC is more active than this project and it's not going anywhere. It works for SBCs now as well as X86 just as it always has. LinuxCNC was always geared towards hobbyists but it's powerful enough that OEMs like [https://tormach.com/](url) use it. Besides that, Machinekit started at a time when MCU dev boards were pretty much just 8 bit Arduino's, if you wanted to do something small a Beaglebone with it's special PRUs was an awesome idea. Now you have 32 bit 600mhz Teensy's for $25 and since LinuxCNC will run on an RPI4, yeah lol....of course Machinekit would die it's a spinoff that's no longer necessary. Mesa even makes RPI4 specific cards

You actually gave me a great idea though, a Raspberry PI CM4  and a Teensy 4.1 stuck to the same board would make an excellent all in one motion controller. Again that has nothing to do with what I'm saying. Some breakout board with a Teensy on it like the ones people already use with grblHAL is the equivalent of a Mesa card, and it is reasonably easy for someone with some electronics knowledge to make a breakout board for a Teensy just as the ones mentioned in the controller repo here are. That is the point. grblHAL obviously doesn't quite work this way but the architecture of the project is already very similar to Mesa's hostmot2 firmware where the stepgens, encoder counter etc. are modules. I've seen it mentioned that grblHAL was actually modeled after LinuxCNC's HAL implementation. grblHAL looks to support quite a few MCU's. I'd think if grblHAL had a LinuxCNC compatabillity you would get alot more people from that side working on this......grblHAL seems to support quite a few MCU's and that's pretty appealing since MCU dev buards are now powerful enough to be taken seriously.

**Reply by terjeio** [reply]
> I don't know how many developers are on this...

On the programming side I am nearly alone - there are a few contributing code for the drivers and plugins.

> You actually gave me a great idea though, a Raspberry PI CM4 and a Teensy 4.1 stuck to the same board would make an excellent all in one motion controller.

Sure is, [check this out](https://github.com/Expatria-Technologies/grblhal_2000_PrintNC). If you look closely you may spot the Pi header.

>  I've seen it mentioned that grblHAL was actually modeled after LinuxCNC's HAL implementation.

This is not correct, the HAL implementation is my idea - it uses function pointers for the HAL quite extensively. If modeled on anything that would be my earlier work with 6502 and ARM processors running RiscOS. 
I mainly use LinuxCNC specifications for extensions above NIST RS274NGC v3. Grbl, which grblHAL is forked from, also did that.

**Reply by ShadeTechnik** [reply]
> > I don't know how many developers are on this...
> 
> On the programming side I am nearly alone - there are a few contributing code for the drivers and plugins.
> 
> > You actually gave me a great idea though, a Raspberry PI CM4 and a Teensy 4.1 stuck to the same board would make an excellent all in one motion controller.
> 
> Sure is, [check this out](https://github.com/Expatria-Technologies/grblhal_2000_PrintNC). If you look closely you may spot the Pi header.
> 
> > I've seen it mentioned that grblHAL was actually modeled after LinuxCNC's HAL implementation.
> 
> This is not correct, the HAL implementation is my idea - it uses function pointers for the HAL quite extensively. If modeled on anything that would be my earlier work with 6502 and ARM processors running RiscOS. I mainly use LinuxCNC specifications for extensions above NIST RS274NGC v3. Grbl, which grblHAL is forked from, also did that.

Honestly I don't know much about grbl or grblHAL other than just looking through the repo quick. Maybe you can assess what might be necessary to make grblHAL working under LinuxCNC similar to as if grblHAL were running like a Mesa card? That would be clearer to someone who may be interested in helping than other parts of this discussion. Some things about lbp protocol are discussed in the thread I linked at the top. Other than that I'd assume LinuxCNC would need a driver/component similar to hm2_eth [https://github.com/LinuxCNC/linuxcnc/tree/master/src/hal/drivers/mesa-hostmot2](url). I don't know how the grblHAL ethernet interface works but  maybe the LinuxCNC side driver could be made compatible with it, otherwise the Mesa Register protocol is described in the manual [https://www.welectron.com/mediafiles/manuals/mesa-electronics/Mesa-Electronics_7i76E_Manual.pdf](url). The grblHAL modules would have to present HAL "pins" to work in LinuxCNC.

**Reply by ShadeTechnik** [reply]
> > You actually gave me a great idea though, a Raspberry PI CM4 and a Teensy 4.1 stuck to the same board would make an excellent all in one motion controller.
> 
> Sure is, [check this out](https://github.com/Expatria-Technologies/grblhal_2000_PrintNC). If you look closely you may spot the Pi header.

I looked at the link quick, It appears that the header is just GPIO and a UART. I'm actually talking about running the realtime interface across the BOB, albeit ethernet or more likely SPI. This is actually done by Mesa with SPI [http://store.mesanet.com/index.php?route=product/product&path=69_62&product_id=345](url) Difference obviously being the FPGA is replaced with a Teensy and I'd use a CM4 myself

**Reply by kfmut** [reply]
> The project you're referring to as "LinuxCNC for Hobbyist" that is in "shambles" for "SBCs" is Machinekit

I was referring to https://forum.linuxcnc.org/18-computer/39037-linuxcnc-orange-pi-allwincnc and other attempts and not to Machinekit at all.

**Reply by ShadeTechnik** [reply]
What on earth does that have to do with anything? For one that's a project for an SBC that I didn't mention until well after you and for 2 it's on hold because you can't even get an Opi, just check the last page of the thread.

A more legit comparison is Remora 
https://github.com/scottalford75/Remora

Basically turns a Smoothieboard or an STM32 into a "mesa card". That's the same Idea, I just don't have any interest in those in particular.

**Reply by terjeio** [reply]
> Maybe you can assess what might be necessary to make grblHAL working under LinuxCNC similar to as if grblHAL were running like a Mesa card?

[This comment](https://github.com/grblHAL/core/discussions/98#discussioncomment-1770783) is still valid. You will have implement the LinuxCNC API/protocol on top of those. If it is possible to reuse the grblHAL HAL a more or less generic version of the LinuxCNC API could be written that could then be usable on many of the grblHAL supported processors.

Personally I am more interested in adding support for the LinuxCNC API to the grblHAL core so that cards using that can be made available. And perhaps the API specs can be used for dual core processors where a core is dedicated to step generation?

**Reply by karoria** [reply]
There are several more gcode commands that are supported by LinuxCNC like cutter radius compensation (G41/G42) which is super useful for some applications. grbl doesn't support it yet. So, I will be interested if this functionality is added to grblHAL by any means.

**Reply by ShadeTechnik** [reply]
> 
> Personally I am more interested in adding support for the LinuxCNC API to the grblHAL core so that cards using that can be made available. And perhaps the API specs can be used for dual core processors where a core is dedicated to step generation?

I can't think of any MCU boards that LinuxCNC supports that are beyond anything of the MCU's you already support in grblHAL. MCU's are just starting to get attention rather than just being tinkered with in LinuxCNC. Pretty hard to justify an MCU over an FPGA for anything but a smaller machine, but it's starting to get there. I think people running LinuxCNC would be alot more likely to use the stuff you already support if it were available to LinuxCNC rather than the other way around

**Reply by terjeio** [reply]
> Pretty hard to justify an MCU over an FPGA for anything but a smaller machine, but it's starting to get there.

I do not get this - bigger machines need more processing power? At what size does this become evident?

> I think people running LinuxCNC would be alot more likely to use the stuff you already support if it were available to LinuxCNC rather than the other way around

I have spent a lot of time developing grblHAL from Grbl, and also on ioSender, leaving my CNC machines largely unused. Adding yet another programming project that will keep my machines idle for way longer is out of the question.
However, I _may_ consider it if I am paid as a professional programmer - at my current rate it will not be cheap.

**Reply by ShadeTechnik** [reply]
> > Pretty hard to justify an MCU over an FPGA for anything but a smaller machine, but it's starting to get there.
> 
> I do not get this - bigger machines need more processing power? At what size does this become evident?

I didn't say "processing power" it's more about cost. A $200 FPGA card (+PC cost since LinuxCNC does not put the interpreter on the controller like grbl) is alot easier to justify in a $5000+ mill than it is in a $600 3D printer. I can't imagine a scenario where an MCU is better than an FPGA since cores run in parallel but I can think of plenty of scenarios where an FPGA is overkill. Personally I have a lot of stuff where a Mesa card is completely unnecessary and I just like to do custom stuff.

> 
> I have spent a lot of time developing grblHAL from Grbl, and also on ioSender, leaving my CNC machines largely unused. Adding yet another programming project that will keep my machines idle for way longer is out of the question. However, I _may_ consider it if I am paid as a professional programmer - at my current rate it will not be cheap.

Well like I said I didn't realize you were the only programmer on this. It's a good open source project, I'm sure a lot of people are grateful you did this. Like I said I'm personally only interested in the Teensy 4 stuff so if it were something I was looking to pay I'd skip grblHAL and get someone to port this [https://github.com/SebKuzminsky/linuxcnc-hal-teensy](url) to a Teensy 4, maybe I'll go that road if necessary. This just seemed like it would be a really cool sort of merger.

**Reply by kfmut** [reply]
> I do not get this - bigger machines need more processing power

It's not about processing power per se, it's about available inputs/outputs, encoders, analog outputs for servo drives and spindle. Modular and highly configurable solutions like Mesa cards(clones included) is clear win here and all those dev boards which are basically done after 4 axes and bunch of inputs are definetly not. LinuxCNC is mainly used to retrofit older commercial equipment and some entry-level new commercial equipment.
